raster_analysis/tiling.py

```diff
@@ -1,144 +1,143 @@
-        if col_name not in groupby_columns:
-            arr = np.array(col_data)
-            merged_tile_results[col_name] = np.bincount(inv, weights=col_data).astype(
-                arr.dtype
+    if not groupby_columns:
+        dataframes = [pd.DataFrame(result, index=[0]) for result in tile_results]
+        merged_df: pd.DataFrame = pd.concat(dataframes)
+        return merged_df.sum().to_dict()
+
+    dataframes = [pd.DataFrame(result) for result in tile_results]
+    merged_df: pd.DataFrame = pd.concat(dataframes)
+
+    grouped_df: pd.DataFrame = merged_df.groupby(groupby_columns).sum()
+    result_df: pd.DataFrame = grouped_df.sort_values(groupby_columns).reset_index()
+
+    # convert ordinal dates to readable dates
+    for col in groupby_columns:
+        if "__date" in col:
+            result_df[col] = result_df[col].apply(
+                lambda val: date.fromordinal(val).strftime("%Y-%m-%d")
+            )
+        elif "__isoweek" in col:
+            result_df[col.replace("__isoweek", "__year")] = result_df[col].apply(
+                lambda val: date.fromordinal(val).isocalendar()[0]
+            )
+            result_df[col] = result_df[col].apply(
+                lambda val: date.fromordinal(val).isocalendar()[1]
             )
 
-    for col, arr in merged_tile_results.items():
-        merged_tile_results[col] = arr.tolist()
-
-    return merged_tile_results
+    return result_df.to_dict(orient="records")
 
 
-def concat_tile_results(tile_results):
-    result = deepcopy(tile_results[0])
+def process_tiled_geoms(tiles, geoprocessing_params, request_id, fanout_num):
+    geom_count = len(tiles)
+    geoprocessing_params["write_to_dynamo"] = True
+    geoprocessing_params["analysis_id"] = request_id
+    logger.info(f"Processing {geom_count} tiles")
 
-    for col in result.keys():
-        for table in tile_results[1:]:
-            result[col] += table[col]
+    tile_geojsons = [mapping(tile) for tile in tiles]
+    tile_chunks = [
+        tile_geojsons[x : x + fanout_num]
+        for x in range(0, len(tile_geojsons), fanout_num)
+    ]
+    fanout_lambda = os.environ["FANOUT_LAMBDA"]
 
-    return result
+    from raster_analysis.boto import invoke_lambda
 
+    for chunk in tile_chunks:
+        event = {"payload": geoprocessing_params, "tiles": chunk}
+        invoke_lambda(event, fanout_lambda, lambda_client())
 
-@xray_recorder.capture("Process Tiled Geometries")
-def process_tiled_geoms(tiled_geoms, geoprocessing_params):
-    execution_threads = []
-    result_queue = queue.Queue()
-    error_queue = queue.Queue()
+    curr_count = 0
+    tries = 0
 
-    for geom in tiled_geoms:
-        payload = geoprocessing_params.copy()
-        payload["geometry"] = mapping(geom)
+    tiled_table = dynamodb_resource().Table(os.environ["TILED_RESULTS_TABLE_NAME"])
+    logger.info(f"Geom count: {geom_count}")
+    while curr_count < geom_count and tries < 60:
+        sleep(0.5)
```

Comment:
makes tries and sleep globals?

I guess there is no better way to let other lambda functions report back to controller lambda function?

Reply:
I think that's what I concluded when I previously researched this. There was no good way to just trigger a new lambda once all the other ones complete to merge the results. You can do it with in step functions with a Map state, but those have a limit of 50 concurrent lambdas.